### ISSUE TITLE
Celo validator gets stuck during initial setup and might require multiple restarts

### ISSUE THREAD

**Issue by zviadm** [celo-org/celo-blockchain#829]
### Expected Behavior

When creating new validator and proxy nodes, for both to start up and catch up on blocks without any intervention.

### Current Behavior

During initial phase of setup, celo-validator can get stuck and not make any progress unless restarted with following messages in the log:
```INFO [01-24|02:26:44.879] Reset round change timer                 address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 func=resetRoundChangeTimer cur_seq=30061 cur_round=0  desired_round=2  state="Waiting for new round" address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 timeout_ms=7µs
INFO [01-24|02:26:51.880] Reset round change timer                 address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 func=resetRoundChangeTimer cur_seq=30061 cur_round=0  desired_round=3  state="Waiting for new round" address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 timeout_ms=11µs
INFO [01-24|02:27:02.881] Reset round change timer                 address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 func=resetRoundChangeTimer cur_seq=30061 cur_round=0  desired_round=4  state="Waiting for new round" address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 timeout_ms=19µs
INFO [01-24|02:27:21.882] Reset round change timer                 address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 func=resetRoundChangeTimer cur_seq=30061 cur_round=0  desired_round=5  state="Waiting for new round" address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 timeout_ms=35µs
INFO [01-24|02:27:56.883] Reset round change timer                 address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 func=resetRoundChangeTimer cur_seq=30061 cur_round=0  desired_round=6  state="Waiting for new round" address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 timeout_ms=67µs
INFO [01-24|02:29:03.884] Reset round change timer                 address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 func=resetRoundChangeTimer cur_seq=30061 cur_round=0  desired_round=7  state="Waiting for new round" address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 timeout_ms=131µs
INFO [01-24|02:31:14.886] Reset round change timer                 address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 func=resetRoundChangeTimer cur_seq=30061 cur_round=0  desired_round=8  state="Waiting for new round" address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 timeout_ms=259µs
INFO [01-24|02:35:33.887] Reset round change timer                 address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 func=resetRoundChangeTimer cur_seq=30061 cur_round=0  desired_round=9  state="Waiting for new round" address=0xfE67E31B487614b02fb18012eEc666A8ff8dCC21 timeout_ms=515µs
```

### Steps to Reproduce Behavior

Setup new validator and proxy on baklava network. Proxy will work just fine and catch up on blocks. Validator will get stuck few times in the beginning.

### Logs

Pasted up above.

### System information

CELO_IMAGE=us.gcr.io/celo-testnet/celo-node:baklava

**Comment by zviadm** [comment]
I was able to reproduce this again on the baklava network. New validator got stuck at two different points in time while syncing initial blocks:
```
INFO [01-25|12:13:35.799] Imported new chain segment               blocks=304 txs=22 mgas=4.433 elapsed=8.020s   mgasps=0.553 number=12358 hash=45a1aa…793285 age=1w6d20h  cache=0.00B
INFO [01-25|12:13:43.201] Imported new chain segment               blocks=285 txs=13 mgas=1.795 elapsed=7.402s   mgasps=0.242 number=12643 hash=08c841…1aad51 age=1w6d19h  cache=0.00B
INFO [01-25|12:13:43.221] Commit new mining work                   number=12644 sealhash=d87701…b3331a uncles=0 txs=0  gas=0       fees=0 elapsed=16.241ms
WARN [01-25|12:13:43.221] Block sealing failed                     err="not an elected validator"
INFO [01-25|12:13:56.213] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=12644 cur_round=0  desired_round=2  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=7µs
INFO [01-25|12:14:03.214] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=12644 cur_round=0  desired_round=3  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=11µs
INFO [01-25|12:14:14.215] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=12644 cur_round=0  desired_round=4  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=19µs
INFO [01-25|12:14:33.216] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=12644 cur_round=0  desired_round=5  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=35µs
INFO [01-25|12:15:08.217] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=12644 cur_round=0  desired_round=6  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=67µs
INFO [01-25|12:16:15.218] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=12644 cur_round=0  desired_round=7  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=131µs
INFO [01-25|12:18:26.219] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=12644 cur_round=0  desired_round=8  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=259µs
INFO [01-25|12:22:45.220] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=12644 cur_round=0  desired_round=9  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=515µs
```

and second time coincided with an epoch change in live network
```
INFO [01-25|12:48:01.635] Imported new chain segment               blocks=235 txs=3183 mgas=195.148 elapsed=8.004s   mgasps=24.378 number=53043 hash=bdf279…78c5d6 age=1w4d12h cache=0.00B
INFO [01-25|12:48:09.659] Imported new chain segment               blocks=231 txs=3084 mgas=186.650 elapsed=8.023s   mgasps=23.262 number=53274 hash=3b5ea6…56860d age=1w4d11h cache=0.00B
ERROR[01-25|12:48:09.860] Error in calling the EVM                 funcName=updateTargetVotingYield transactionData=0x92ecd745 err="evm: execution reverted"
ERROR[01-25|12:48:09.860] Error when invoking evm function         err="evm: execution reverted"                                         funcName=updateTargetVotingYield static=false address=0x188A97403db355DB1968d039aE9C1Db2c33663F7 args=[] gas=2000000 gasLeft=1997356 value=0
ERROR[01-25|12:48:09.860] Error in executing function on registered contract function=updateTargetVotingYield registryId=0xb33c3d77234979a288baf651b42d19d8483d5af3e328f9f1fae2bef7b11acd25 err="evm: execution reverted"
INFO [01-25|12:48:17.692] Imported new chain segment               blocks=231 txs=3176 mgas=190.274 elapsed=8.032s   mgasps=23.688 number=53505 hash=558850…2ad543 age=1w4d11h cache=0.00B
INFO [01-25|12:48:19.529] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=52194 cur_round=0  desired_round=5  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=35µs
INFO [01-25|12:48:25.704] Imported new chain segment               blocks=246 txs=3227 mgas=196.505 elapsed=8.012s   mgasps=24.525 number=53751 hash=f5ca73…7069fc age=1w4d11h cache=0.00B
INFO [01-25|12:48:28.305] Imported new chain segment               blocks=82  txs=1026 mgas=62.448  elapsed=2.600s   mgasps=24.009 number=53833 hash=f91ba4…2772ad age=1w4d11h cache=0.00B
INFO [01-25|12:48:28.331] Commit new mining work                   number=53834 sealhash=55aa63…228bb1 uncles=0 txs=0    gas=0       fees=0 elapsed=16.736ms
WARN [01-25|12:48:28.332] Block sealing failed                     err="not an elected validator"
INFO [01-25|12:48:41.322] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=53834 cur_round=0  desired_round=2  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=7µs
INFO [01-25|12:48:48.323] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=53834 cur_round=0  desired_round=3  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=11µs
INFO [01-25|12:48:59.324] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=53834 cur_round=0  desired_round=4  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=19µs
INFO [01-25|12:49:18.325] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=53834 cur_round=0  desired_round=5  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=35µs
INFO [01-25|12:49:53.326] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=53834 cur_round=0  desired_round=6  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=67µs
INFO [01-25|12:51:00.327] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=53834 cur_round=0  desired_round=7  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=131µs
INFO [01-25|12:53:11.328] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=53834 cur_round=0  desired_round=8  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=259µs
INFO [01-25|12:57:30.329] Reset round change timer                 address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 func=resetRoundChangeTimer cur_seq=53834 cur_round=0  desired_round=9  state="Waiting for new round" address=0xAfcEfa1e198C0f486e0700eDd014Ed33967Bf3b8 timeout_ms=515µs
thezviad_gmail_com@cv-n20200125-2:~$ sudo docker restart celo-validator
```

**Comment by zviadm** [comment]
is this potentially related to the election issues that was on baklava network? since election contract got updated in later blocks to fix the bug, could it be that now it requires multiple restarts to get through the initial blocks?

**Comment by nategraf** [comment]
I believe this is related to https://github.com/celo-org/celo-blockchain/pull/840. It's hard to know for sure from those logs, but the symptoms all line up.

**Comment by zviadm** [comment]
> I believe this is related to #840. It's hard to know for sure from those logs, but the symptoms all line up.

this is very easy to reproduce on baklava right now, so let me know if you want me to run new validator with specific flags and i can give you more detailed logs. (or if you have a potential fix I can test it out too)

**Comment by nategraf** [comment]
If you are inclined to try it out, your could pull `us.gcr.io/celo-testnet/geth:e48cb9d7113916b0aafa6b8ab39baa1a23693847` which is built from https://github.com/celo-org/celo-blockchain/commit/e48cb9d7113916b0aafa6b8ab39baa1a23693847 and give it a go. Otherwise we'll try to get to this soon, but we have a bit of a backlog.

**Comment by zviadm** [comment]
> If you are inclined to try it out, your could pull `us.gcr.io/celo-testnet/geth:e48cb9d7113916b0aafa6b8ab39baa1a23693847` which is built from [e48cb9d](https://github.com/celo-org/celo-blockchain/commit/e48cb9d7113916b0aafa6b8ab39baa1a23693847) and give it a go. Otherwise we'll try to get to this soon, but we have a bit of a backlog.
> 
> Note: This is the same as the `celo-node` image, but does not come packaged with the genesis block or bootnodes file, so you'll want to pull those from you existing setup or the `us.gcr.io/celo-testnet/celo-node:baklava` image.

I am assuming  https://github.com/celo-org/celo-blockchain/commit/e48cb9d7113916b0aafa6b8ab39baa1a23693847 will be in new image that will be released on Monday so it will be easier to try that one out directly.

**Comment by zviadm** [comment]
btw I grabbed a stack dump from stuck validator at startup:

[stuckvalidator_stacks.txt](https://github.com/celo-org/celo-blockchain/files/4156930/stuckvalidator_stacks.txt)

**Comment by zviadm** [comment]
some more debugging info. interestingly just restarting `celo-proxy` node also  unblocks the validator with following log lines:
```
INFO [02-05|01:33:03.456] Reset round change timer                 address=0xe30065c6a165F5D06583e69590c9f2008cf93c75 func=resetRoundChangeTimer cur_seq=14776 cur_round=0  desired_round=11 state="Waiting for new round" address=0xe30065c6a165F5D06583e69590c9f2008cf93c75 timeout_ms=2.051ms
INFO [02-05|02:04:19.547] Ethereum peer connected                  id=968811a5cf7845ed conn=staticdial name=Geth/v1.8.23-stable/linux-amd64/go1.11.13
WARN [02-05|02:04:54.297] Synchronisation failed, retrying         err="block download canceled (requested)"
ERROR[02-05|02:05:22.555] No proxy peers, cannot send Istanbul Validator Enodes Share message 
INFO [02-05|02:05:24.309] Ethereum peer connected                  id=968811a5cf7845ed conn=staticdial name=Geth/v1.8.23-stable/linux-amd64/go1.11.13
INFO [02-05|02:06:13.407] Imported new chain segment               blocks=2   txs=0   mgas=0.000  elapsed=63.634ms  mgasps=0.000  number=14777 hash=d24f15…12da7d age=3w3d6h  cache=0.00B
INFO [02-05|02:06:13.423] Commit new mining work                   number=14778 sealhash=6ba74b…901339 uncles=0 txs=0   gas=0       fees=0 elapsed=15.794ms
WARN [02-05|02:06:13.423] Block sealing failed                     err="not an elected validator"
INFO [02-05|02:06:16.954] Imported new chain segment               blocks=113 txs=2   mgas=0.620  elapsed=3.540s    mgasps=0.175  number=14890 hash=9d3789…a914e4 age=3w3d6h  cache=0.00B
INFO [02-05|02:06:16.968] Commit new mining work                   number=14891 sealhash=c6121f…1cc2e7 uncles=0 txs=0   gas=0       fees=0 elapsed=13.882ms
WARN [02-05|02:06:16.968] Block sealing failed                     err="not an elected validator"
```

and it continues  start syncing blocks just fine.

**Comment by nambrot** [comment]
Look at the stacks, indeed it is not the one solved with #840 . There are some eerie parallels to it though, notably the reconnection allowing us to sync blocks again. If you run into this again, if you can enable TRACE logging via `--verbosity 5` or `debug.verbosity(5)` in the console

**Comment by zviadm** [comment]
this is most likely resolved with 1.9, i havent seen this occur in mainnet or new baklava at all.
